server/src/main/java/com/linecorp/line/auth/fido/fido2/server/util/MdsV3MetadataCertificateUtil.java

```diff
@@ -0,0 +1,203 @@
+        Set<TrustAnchor> trustAnchors = CertificateUtil.getTrustAnchors(mdsInfo.getRootCertificates());
+
+        try {
+            boolean result = CertPathUtil.validate(certificateChain, trustAnchors, true);
+
+            if (!result) {
+                throw new MdsV3MetadataException(MetadataTOCResult
+                        .builder()
+                        .result(false)
+                        .totalCount(metadataBLOBPayload.getEntries().size())
+                        .updatedCount(0)
+                        .reason("Chain validation fail")
+                        .build());
+            }
+        } catch (GeneralSecurityException e) {
+            throw new MdsV3MetadataException(MetadataTOCResult
+                    .builder()
+                    .result(false)
+                    .totalCount(metadataBLOBPayload.getEntries().size())
+                    .updatedCount(0)
+                    .reason("Chain validation exception: " + e.getMessage())
+                    .build());
+        }
+
+        return certificateChain;
+    }
+
+    private static List<Certificate> getCertificateChain(String url, List<String> rootCertificates, DecodedJWT decodedJWT, MetadataBLOBPayload metadataBLOBPayload) throws CertificateException, MdsV3MetadataException {
+        if (!decodedJWT.getHeaderClaim("x5u").isNull()) {
+            return getX5UCertificates(url, decodedJWT, metadataBLOBPayload);
+        } else {
+            return getX5CCertificates(rootCertificates, decodedJWT, metadataBLOBPayload);
+        }
+    }
+
+    private static List<Certificate> getX5UCertificates(String url, DecodedJWT decodedJWT, MetadataBLOBPayload metadataBLOBPayload) throws CertificateException, MdsV3MetadataException {
+        // refer x5u
+        List<Certificate> certificateChain = new ArrayList<>();
+        String x509UrlString = decodedJWT.getHeaderClaim("x5u").asString();
+
+        URL x509Url, tocUrl;
+        try {
+            x509Url = new URL(x509UrlString);
+            tocUrl = new URL(url);
+
+        } catch (MalformedURLException e) {
+            throw new MdsV3MetadataException(MetadataTOCResult
+                    .builder()
+                    .result(false)
+                    .totalCount(metadataBLOBPayload.getEntries().size())
+                    .updatedCount(0)
+                    .reason("URL for x5u is not valid")
+                    .build());
+        }
+
+        if (!x509Url.getHost().equals(tocUrl.getHost())) {
+            throw new MdsV3MetadataException(MetadataTOCResult
+                    .builder()
+                    .result(false)
+                    .totalCount(metadataBLOBPayload.getEntries().size())
+                    .updatedCount(0)
+                    .reason("x5u origin differs to Metadata TOC origin")
+                    .build());
+        }
+
+        // retrieve x509 certificate or certificate chain (PEM)
+        String pemEncoded = ""; // need to get from x5u url
```

Comment:
My understanding is that `x5u` signature is not supported yet.
Maybe we should throw an exception saying it is not supported.